src/main/kotlin/dev/supergrecko/kllvm/core/typedefs/Type.kt

```diff
@@ -59,50 +64,34 @@ public open class Type internal constructor(
             require(!isOpaque())
         }
 
-        return Value(LLVM.LLVMConstNull(llvmType))
+        return Value(LLVM.LLVMConstNull(ref))
     }
 
     /**
      * @see [LLVM.LLVMGetUndef]
      */
     public fun getConstantUndef(): Value {
-        return Value(LLVM.LLVMGetUndef(llvmType))
+        return Value(LLVM.LLVMGetUndef(ref))
     }
 
     /**
      * @see [LLVM.LLVMConstPointerNull]
      */
     public fun getConstantNullPointer(): Value {
-        return Value(LLVM.LLVMConstPointerNull(llvmType))
+        return Value(LLVM.LLVMConstPointerNull(ref))
     }
     //endregion Core::Values::Constants
 
     //region Typecasting
-    public fun toPointerType(addressSpace: Int = 0): PointerType = PointerType.new(this, addressSpace)
+    public fun toPointerType(addressSpace: Int = 0): PointerType =
+        PointerType(this, addressSpace)
 
-    public fun toArrayType(size: Int): ArrayType = ArrayType.new(this, size)
-
-    public fun toVectorType(size: Int): VectorType = VectorType.new(this, size)
-
-    public inline fun <reified T : Type> cast(): T {
-        val ctor: Constructor<T> = T::class.java.getDeclaredConstructor(LLVMTypeRef::class.java)
-
-        return ctor.newInstance(getUnderlyingReference())
-                // Should theoretically be unreachable
-                ?: throw TypeCastException("Failed to cast LLVMType to T")
-    }
+    public fun toArrayType(size: Int): ArrayType = ArrayType(this, size)
 
-    public fun asArrayType(): ArrayType = ArrayType(llvmType)
-    public fun asFloatType(): FloatType = FloatType(llvmType)
-    public fun asFunctionType(): FunctionType = FunctionType(llvmType)
-    public fun asIntType(): IntType = IntType(llvmType)
-    public fun asPointerType(): PointerType = PointerType(llvmType)
-    public fun asStructType(): StructType = StructType(llvmType)
-    public fun asVectorType(): VectorType = VectorType(llvmType)
-    public fun asVoidType(): VoidType = VoidType(llvmType)
```

Comment:
This is now redundant as all the Type wrapper classes have a constructor like this:

```kotlin
public constructor(llvmType: LLVMTypeRef) : this() {
    ref = llvmType
}
```

Which means we're able to do unsafe casts like this now:

```kotlin
val ty = VoidType()
val int = IntType(ty) // unsafe and unreasonable cast, but it offers the same API as before

Reply:
You mean 
```
val ty = VoidType()
val int = IntType(ty.getUnderlyingReference()) // unsafe and unreasonable cast, but it offers the same API as before
```
or am I missing something?

Reply:
My bad, copied the wrong constructor :P

This is what I was thinking of

```kotlin
public constructor(type: Type) : this(type.ref)
```